src/Magnum/Platform/EmscriptenApplication.cpp

```diff
@@ -0,0 +1,474 @@
+    DEALINGS IN THE SOFTWARE.
+*/
+
+#include "EmscriptenApplication.h"
+
+#include <emscripten/emscripten.h>
+
+#include <Corrade/Containers/ArrayView.h>
+#include <Corrade/Utility/Arguments.h>
+#include <Corrade/Utility/Debug.h>
+#include <Corrade/Utility/String.h>
+
+#include "Magnum/Math/ConfigurationValue.h"
+
+#include "Magnum/GL/Version.h"
+#include "Magnum/Platform/GLContext.h"
+
+#include "Magnum/Platform/Implementation/DpiScaling.h"
+
+namespace Magnum { namespace Platform {
+
+namespace Implementation {
+    typedef EmscriptenApplication::KeyEvent::Key Key;
+
+    /* Entry for key name to `Key` enum mapping */
+    struct Entry {
+        const char* key;
+        Key value;
+    };
+
+    /* Key name to `Key` enum mapping. These need to be sorted for use with
+       std::lower_bound in KeyEvent::toKey */
+    constexpr Entry KEY_MAPPING[]{
+        {"AltLeft", Key::LeftAlt},
+        {"AltRight", Key::RightAlt},
+        {"ArrowDown", Key::Down},
+        {"ArrowLeft", Key::Left},
+        {"ArrowRight", Key::Right},
+        {"ArrowUp", Key::Up},
+        {"Backslash", Key::Backslash},
+        {"Backspace", Key::Backspace},
+        {"CapsLock", Key::CapsLock},
+        {"Comma", Key::Comma},
+        {"ContextMenu", Key::Menu},
+        {"ControlLeft", Key::LeftCtrl},
+        {"ControlRight", Key::RightCtrl},
+        {"Delete", Key::Delete},
+        {"End", Key::End},
+        {"Enter", Key::Enter},
+        {"Equal", Key::Equal},
+        {"Escape", Key::Esc},
+        {"Home", Key::Home},
+        {"Insert", Key::Insert},
+        {"MetaLeft", Key::LeftSuper},
+        {"MetaRight", Key::RightSuper},
+        {"Minus", Key::Minus},
+        {"NumLock", Key::NumLock},
+        {"PageDown", Key::PageDown},
+        {"PageUp", Key::PageUp},
+        {"Pause", Key::Pause},
+        {"Period", Key::Period},
+        {"Plus", Key::Plus},
+        {"PrintScreen", Key::PrintScreen},
+        {"Quote", Key::Quote},
+        {"ScrollLock", Key::ScrollLock},
+        {"ShiftLeft", Key::LeftShift},
+        {"ShiftRight", Key::RightShift},
+        {"Slash", Key::Slash},
+        {"Space", Key::Space},
+        {"Tab", Key::Tab},
+    };
+
+    /* Predicate for Entry "less than" to use with std::lower_bound */
+    struct EntryCompare {
+        bool operator()(const Entry& a, const char* const& b) {
+            return std::strcmp(a.key, b) < 0;
+        }
+
+        bool operator()(const char*& a, const Entry& b) {
+            return std::strcmp(a, b.key) < 0;
+        }
+    };
+
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wmissing-prototypes"
+    /* Returns the id of the element that is used for listening to key events
+       or '#document' if none is configured in the Module */
+    EM_JS(const char*, getKeyboardListeningElement, (), {
+        return Module['keyboardListeningElement'] || '#document';
+    });
+    #pragma GCC diagnostic pop
```

Comment:
> wasm symbol export and 

There is no symbol exported from wasm, though...

As far as I know, this generates basically the same JavaScript/WASM code, but with a temporary unnamed function. I therefore would very much prefer the function as you can read the code in a "default-C++-reading mode" until you actually look up the implementation of `getKeyboardListeningElement`. Your version is pretty hard to read, but I fear that mine returns memory ownership and `const char*` is therefore not correct. OTOH... `intArrayFromString` sounds like it will have 3 bytes inbetween the chars...?

"inlining JS" in this way only inlines in the C++ code, not in the binary.

I will compile both and report back when I verified that this is the case and know what's going on with those allocations :)

Reply:
Summarizing follow-up discussion on gitter for later reference:

`EM_ASM` produces:

~~~js
var ASM_CONSTS = [function() { var element = Module['keyboardListeningElement'] || '#document'; return allocate(intArrayFromString(element), 'i8', ALLOC_NORMAL); },
 function($0) { var name = UTF8ToString($0); return typeof process !== 'undefined' && name in process.env ? allocate(intArrayFromString(process.env[name]), 'i8', ALLOC_NORMAL) : 0; }];

function _emscripten_asm_const_i(code) {
  return ASM_CONSTS[code]();
}

function _emscripten_asm_const_ii(code, a0) {
  return ASM_CONSTS[code](a0);
}
~~~

`EM_JS` produces:

~~~js
var ASM_CONSTS = [function($0) { var name = UTF8ToString($0); return typeof process !== 'undefined' && name in process.env ? allocate(intArrayFromString(process.env[name]), 'i8', ALLOC_NORMAL) : 0; }];

function _emscripten_asm_const_ii(code, a0) {
  return ASM_CONSTS[code](a0);
}
function _getKeyboardListeningElement(){ return Module['keyboardListeningElement'] || '#document'; }
~~~

... on emscripten version `1.38.6`.

While the former seems to produce more boilerplate code around the function call, there are other functions in corrade that would generate those also. 

The `EM_ASM` version makes the necessity of the `std::free` way clearer. In the `EM_JS` version this is not apparent and I actually forgot the `std::free` here. We decided to use `EM_ASM` for the time being.